Automattic/newspack-network
# Epic/subscription membership sync

This epic refactors the strategy to check for users' subscriptions across the network, as well as recovering the Woo centralized data based on sync events.

I did this on a series of chained PRs to make it easier to parse. They need to be reviewed and merged in order into the Epic branch. Last step is to merge this epic branch into trunk.

In this epic branch, we revert https://github.com/Automattic/newspack-network/pull/114, https://github.com/Automattic/newspack-network/commit/cb5bcb7be0235c72a21d89ffcdac6ae154a3fa15 and https://github.com/Automattic/newspack-network/commit/e23442525922b95d66455ff48b40b1cb85ed7eeb, and do a little fix to ensure the data is always parsed the same way (https://github.com/Automattic/newspack-network/pull/122/commits/9140a8ebe8accd71d1eba19a46f532f10a4d30fa)

## PRs to be merged in order

### 1: https://github.com/Automattic/newspack-network/pull/123

This simply creates backfillers for the existing Woo events and creates a user meta where we store the user's network subscriptions.

Testing:

* On a site with existing Woo data, run `wp newspack-network data-backfill newspack_node_subscription_changed` and `wp newspack-network data-backfill newspack_node_order_changed`. Test it in `verbose` and normal modes
* Make sure the events are created correctly
* Propagate the events in the network and make sure the Central Woo dashboard looks right (use `wp newspack-network process-webhooks` in the Node to speed syncing up)
* Find a user that has a subscription in a site and take note of their email
* Find this user in another site in the network
* Get the usermeta: `get_user_meta( $user_id, '_newspack_network_subscriptions', true);`
* Confirm the user meta contains the subscriptions this user has in other sites in the network

###  2: https://github.com/Automattic/newspack-network/pull/124

This PR creates the `membership_plan_updated` event

Testing

* On a site with existing membership plans, run `wp newspack-network data-backfill newspack_network_membership_plan_updated` 
* Confirm events for all existing membership plans is created
* Propagate the events (either pulling on the Node site or using `process-webhooks` to send events to the hub)
* Check the option on the sites include all existing membership plans in the other sites `get_option( 'newspack_network_membership_plans' );`
* Edit a membership plan in one site
* Confirm it created a new `membership_plan_updated` event
* Propagate the event
* Make sure the option is updated in the target site

## ALTERNATIVE

### https://github.com/Automattic/newspack-network/pull/134

The above PR brings back #114 .

We can merge it and release. Then we'll have all the new events. After they are deployed and all the backfills are executed. We can revert #114 again and proceed with the remaining PRs


## END OF ALTERNATIVE

###  3: https://github.com/Automattic/newspack-network/pull/125

This is a re-implementation of https://github.com/Automattic/newspack-network/pull/114

Testing:

* Create a granted-on-subscription memberships plan on two network sites, and make them synchronised (set same Network ID in the plan settings)
* On the Hub, buy a subscription tied to that membership plan as a reader
* On the Node, Pull all events
* Visit the Node as an anonymous reader and buy a subscription that gives access to the same membership plan using the same email
* Run  `wp newspack-network process-webhooks` in the node to sync events to the Hub
* On the Hub admin, Cancel the subscription
* Confirm the subscription is kept active, and that a note is added to the user membership saying it was kept active because there is an active subscription in the Node

###  4: https://github.com/Automattic/newspack-network/pull/126

This is a re-implementation of https://github.com/Automattic/newspack-network/pull/112

Testing: 

* Create a granted-on-subscription memberships plan on two network sites, and make them synchronised (set same Network ID in the plan settings)
* Make sure to sync the latest events by running  `wp newspack-network process-webhooks` and `wp newspack-network sync-all` in the Node 
* you should see the `newspack_network_membership_plan_updated` events in the Event log for both the plan in the hub and the plan in the node
* you also should see that the `newspack_network_membership_plans` option populated in both sites (this is what we tested in #124 
* One one site, buy the subscription, which should grant you the plan
* On the second site, log in/register using the same email address
* On the second site, try to buy the membership-granting subscription – observe it's not allowed

![Captura de tela de 2024-08-07 12-42-47](https://github.com/user-attachments/assets/341a512c-dbd8-429f-acb6-bc64377c2ff9)
